aggregator/flush_mgr.go

```diff
@@ -0,0 +1,338 @@
+	Flush()
+}
+
+// FlushManager manages and coordinates flushing activities across many
+// periodic flushers with different flush intervals with controlled concurrency
+// for flushes to minimize spikes in CPU load and reduce p99 flush latencies.
+type FlushManager interface {
+	// Register registers a metric list with the flush manager.
+	Register(flusher PeriodicFlusher) error
+
+	// Close closes the flush manager.
+	Close()
+}
+
+type flushManagerMetrics struct {
+	queueSize tally.Gauge
+}
+
+func newFlushManagerMetrics(scope tally.Scope) flushManagerMetrics {
+	return flushManagerMetrics{
+		queueSize: scope.Gauge("queue-size"),
+	}
+}
+
+type flushManager struct {
+	sync.Mutex
+
+	nowFn         clock.NowFn
+	checkEvery    time.Duration
+	jitterEnabled bool
+	workers       xsync.WorkerPool
+	scope         tally.Scope
+
+	closed     bool
+	rand       *rand.Rand
+	buckets    []*flushBucket
+	flushTimes flushMetadataHeap
+	sleepFn    sleepFn
+	wgFlush    sync.WaitGroup
+	metrics    flushManagerMetrics
+}
+
+// NewFlushManager creates a new flush manager.
+func NewFlushManager(opts FlushManagerOptions) FlushManager {
+	if opts == nil {
+		opts = NewFlushManagerOptions()
+	}
+	nowFn := opts.ClockOptions().NowFn()
+	scope := opts.InstrumentOptions().MetricsScope()
+	rand := rand.New(rand.NewSource(nowFn().UnixNano()))
+	workers := xsync.NewWorkerPool(opts.WorkerPoolSize())
+	workers.Init()
+
+	mgr := &flushManager{
+		nowFn:         nowFn,
+		checkEvery:    opts.CheckEvery(),
+		jitterEnabled: opts.JitterEnabled(),
+		workers:       workers,
+		scope:         scope,
+		rand:          rand,
+		sleepFn:       time.Sleep,
+		metrics:       newFlushManagerMetrics(scope),
+	}
+	if mgr.checkEvery > 0 {
+		mgr.wgFlush.Add(1)
+		go mgr.flush()
+	}
+	return mgr
+}
+
+func (mgr *flushManager) Register(flusher PeriodicFlusher) error {
+	mgr.Lock()
+	bucket, err := mgr.bucketForWithLock(flusher)
+	if err == nil {
+		bucket.Add(flusher)
+		mgr.Unlock()
+		return nil
+	}
+	mgr.Unlock()
+	return err
+}
+
+func (mgr *flushManager) bucketForWithLock(l PeriodicFlusher) (*flushBucket, error) {
+	if mgr.closed {
+		return nil, errFlushManagerClosed
+	}
+	flushInterval := l.FlushInterval()
+	for _, bucket := range mgr.buckets {
+		if bucket.interval == flushInterval {
+			return bucket, nil
+		}
+	}
+	bucketScope := mgr.scope.SubScope("bucket").Tagged(map[string]string{
+		"interval": flushInterval.String(),
+	})
+	bucket := newBucket(flushInterval, bucketScope)
+	mgr.buckets = append(mgr.buckets, bucket)
+
+	// NB(xichen): this is a new bucket so we need to update the flush times
+	// heap for the flush goroutine to pick it up.
+	nowNanos := mgr.nowNanos()
+	nextFlushNanos := nowNanos + int64(flushInterval)
+	if mgr.jitterEnabled {
+		jitterNanos := int64(mgr.rand.Int63n(int64(flushInterval)))
+		nextFlushNanos = nowNanos + jitterNanos
+	}
+	newFlushMetadata := flushMetadata{
+		timeNanos: nextFlushNanos,
+		bucketIdx: len(mgr.buckets) - 1,
```

Comment:
why not use a map for buckets keyed on interval? so here can reference to the interval?